[memgraph-2-12] Write docs for merge module (#277)
memgraph/documentation: pages/advanced-algorithms/available-algorithms/merge.mdx

description: Find out how to merge nodes and relationships within the graph structure with more precision and flexibility.
---

import { Callout } from 'nextra/components'
import { Card, Cards } from 'nextra/components'
import GitHub from '/components/icons/GitHub'

# merge

The `merge` module provides the capabilities of the `MERGE` Cypher command, merging or creating nodes and relationships as per specified conditions.
It ensures precision and coherence in managing interconnected data structures.

<Cards>
  <Card
    icon={<GitHub />}
    title="Source code"
    href="https://github.com/memgraph/mage/tree/main/cpp/merge_module"
  />
</Cards>

| Trait               | Value               |
| ------------------- | ------------------- |
| **Module type**     | util                |
| **Implementation**  | C++                 |
| **Graph direction** | directed/undirected |
| **Edge weights**    | weighted/unweighted |
| **Parallelism**     | sequential          |

## Procedures

<Callout type="info">
You can execute this algorithm on [graph projections, subgraphs or portions of the graph](/advanced-algorithms/run-algorithms#run-procedures-on-subgraph).
</Callout>

### `node()`

> Provides a more flexible way of merging nodes than Cypher’s `MERGE` clause, by

Reviewer: ```suggestion
The functions provides a more flexible way of merging nodes than Cypher’s `MERGE` clause, by
```

allowing labels and properties to be extracted as results of other procedures
and sent as a list or map.

{<h4> Input: </h4>}

> - `labels: List[string]` ➡ A list of all the labels to be added to the new node.

Reviewer: ```suggestion
- `labels: List[string]` ➡ A list of all labels that need to be added to the new node.
```

> - `identProps: Map` ➡ A map with key/value pairs used for identifying which nodes will be merged.
> - `createProps: Map` ➡ A map with key/value pairs which represent properties of the node, when the procedure creates a new node.
> Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
> - `mergeProps: Map` ➡ A map with key/value pairs which represent properties of the node, when the procedure merges onto a node.

Reviewer: ```suggestion
- `identProps: Map` ➡ A map with key-value pairs used for identifying which nodes will be merged.
- `createProps: Map` ➡ A map with key-value pairs which represent properties of the node, when the procedure creates a new node.
Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
- `mergeProps: Map` ➡ A map with key-value pairs which represent properties of the node, when the procedure merges onto a node.
```

Equivalent to using `ON MATCH SET` for properties when using `MERGE`.

{<h4> Output: </h4>}

- `node: Node` ➡ The node which was created or merged in the procedure.

> {<h4> Usage: </h4>}

Reviewer: ```suggestion

{<h4> Usage: </h4>}
```

> Consider a graph with two `Human`, and one `Alien` node.

Reviewer: ```suggestion

Consider a graph with two `Human`, and one `Alien` node.
```

```cypher
CREATE (h:Human {name:"Mark", age: 20}), (h2:Human {name:"Eva", age: 30}), (a:Alien {name: "Gorlock", age: 20});
```

> Let's use `merge.node` on all our nodes, and add `galaxy` property on existing nodes. By leaving `labels` and `identProp` empty,
> all nodes in the graph will be affected by the procedure.

Reviewer: ```suggestion
Use the `merge.node` procedure on all nodes, and add `galaxy` property on the existing nodes. By leaving `labels` and `identProp` empty,
all nodes in the graph will be affected by the procedure.
```

```cypher

> CALL merge.node([],{},{},{galaxy: "Milky way"}) YIELD node RETURN node;

Reviewer: ```suggestion
CALL merge.node([],{},{},{galaxy: "Milky way"}) 
YIELD node 
RETURN node;
```

```

```plaintext
+----------------------------------------------------------------------------------------------------------------+
| node                                                                                                           |
+----------------------------------------------------------------------------------------------------------------+
| {"id": 0,"labels": ["Human"],"properties": {"age": 20,"galaxy": "Milky way","name": "Mark"},"type": "node"}    |
+----------------------------------------------------------------------------------------------------------------+
| {"id": 1,"labels": ["Human"],"properties": {"age": 30,"galaxy": "Milky way","name": "Eva"},"type": "node"}     |
+----------------------------------------------------------------------------------------------------------------+
| {"id": 2,"labels": ["Alien"],"properties": {"age": 20,"galaxy": "Milky way","name": "Gorlock"},"type": "node"} |
+----------------------------------------------------------------------------------------------------------------+
```

> Now, let's use `merge.node` to either merge or create an `Alien` node with the name `Zumbi`. If that's a newly created node,
> `age` property will be 30, and if it's a merged node, `age` will be 40. Additionally, `galaxy` property will be used as an identifying property.
> NOTE: When using `merge.node` or `MERGE`, if there is no match in the graph, and a node is created, it will be created with identifying properties
> alongside create properties.

Reviewer: ```suggestion
Now, use the `merge.node` procedure to either merge or create an `Alien` node with the name `Zumbi`. If that's a newly created node, the
`age` property will be 30, and if it's a merged node, `age` will be 40. Additionally, the `galaxy` property will be used as an identifying property.

When using `merge.node` procedure or `MERGE` clause, if there is no match in the graph, and a node is created, it will be created with identifying properties
alongside create properties.
```

I don't understand this -  "it will be created with identifying properties alongside CREATE PROPERTIES? What do you mean with create properties?

Author: In the input section, I defined label, identProp, CreateProp, MergeProp. how merge works is it finds the nodes based on the label and identfying poperties, and sets MergeProperties on that node. If the node doesnt exist, it will create a new node, and set its properties to be the same like identProp and CreateProp(Create Properties). For example, label = Zumbi, identProp = {age =20), createProp = {galaxy: Andromeda}. If there is no node labeled Zumbi with age 20, it will be created, and it will also get property galaxy.

Reviewer: i've added defined by the `createProps` input parameter.

Author: Great!

```cypher
CALL merge.node(["Zumbi"],{galaxy: "Andromeda"},{age:20},{age: 40}) YIELD node RETURN node;
```

```plaintext
+----------------------------------------------------------------------------------------------------------------+
| node                                                                                                           |
+----------------------------------------------------------------------------------------------------------------+
| {"id": 3,"labels": ["Zumbi"],"properties": {"age": 20,"galaxy": "Andromeda"},"type": "node"}                  |
+----------------------------------------------------------------------------------------------------------------+

```

> As we can see, a new node was created, because the graph did not contain an `Alien` node with name `Zumbi` before.

Reviewer: ```suggestion

A new node was created, because the graph did not contain an `Alien` node with the name `Zumbi.
```

> Now, let's use `merge.node` to give weapons to all `Humans` aged 20.

Reviewer: ```suggestion
Now, use the `merge.node` procedure to give weapons to all `Humans` aged 20.
```

```cypher
CALL merge.node(["Zumbi"],{galaxy: "Andromeda"},{age:20},{age: 40}) YIELD node RETURN node;
```

```plaintext
+---------------------------------------------------------------------------------------------------------------------------------+
| node                                                                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------+
| {"id": 0,"labels": ["Human"],"properties": {"age": 20,"galaxy": "Milky way","name": "Mark","weapon": "sword"},"type": "node"}  |
+---------------------------------------------------------------------------------------------------------------------------------+

```

> As we can see, `Mark` is the only node on the graph that can get a `weapon` property.

Reviewer: ```suggestion
`Mark` is the only node on the graph that can get a `weapon` property.
```

+---------------------------------------------------------------------------------------------------------------------------------+
| {"id": 0,"labels": ["Human"],"properties": {"age": 25,"galaxy": "Milky way","name": "Mark","weapon": "sword"},"type": "node"}   |
+---------------------------------------------------------------------------------------------------------------------------------+
| {"id": 2,"labels": ["Alien"],"properties": {"age": 25,"galaxy": "Milky way","name": "Gorlock"},"type": "node"}                  |
+---------------------------------------------------------------------------------------------------------------------------------+
| {"id": 33,"labels": ["Zumbi"],"properties": {"age": 25,"galaxy": "Andromeda"},"type": "node"}                                   |
+---------------------------------------------------------------------------------------------------------------------------------+
```

### `relationship()`

Provides a more flexible way of merging relationships than Cypher’s `MERGE` clause, by
allowing labels and properties to be extracted as results of other procedures
and sent as a list or map.

{<h4> Input: </h4>}

> - `startNode: Node` ➡ Start node of the relationship.
> - `relationshipType: string` ➡ Relationship type to be matched or created.
> - `identProps: Map` ➡ A map with key/value pairs used for identifying which relationships will be merged.
> - `createProps: Map` ➡ A map with key/value pairs which represent properties of the relationship, when the procedure creates a new relationship.
> Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
> - `endNode: Node` ➡ End node of the relationship.
> - `mergeProps: Map` ➡ A map with key/value pairs which represent properties of the relationship, when the procedure merges onto a relationship.
> Equivalent to using `ON MATCH SET` for properties when using `MERGE`.

Reviewer: ```suggestion
- `startNode: Node` ➡ The start node of the relationship.
- `relationshipType: string` ➡ The relationship type to be matched or created.
- `identProps: Map` ➡ A map with key-value pairs used for identifying which relationships will be merged.
- `createProps: Map` ➡ A map with key-value pairs which represent properties of the relationship, when the procedure creates a new relationship.
Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
- `endNode: Node` ➡ The end node of the relationship.
- `mergeProps: Map` ➡ A map with key-value pairs which represent properties of the relationship, when the procedure merges onto a relationship.
Equivalent to using `ON MATCH SET` for properties when using `MERGE`.
```

{<h4> Output: </h4>}

- `rel: Relationship` ➡ The relationship which was created or merged in the procedure.

{<h4> Usage: </h4>}

> Consider a graph with one `Human`, two `Dog` nodes. Both dogs love the human but one loves him unconditionally.

Reviewer: ```suggestion

Consider a graph with one `Human`, two `Dog` nodes. Both dogs love the human but one loves him unconditionally.
```

```cypher

> CREATE (d:Dog {name: 'Johnny'})-[l:Loves {unconditionally: true}]->(h:Human) CREATE (c:Dog {name: 'Floki'});

Reviewer: ```suggestion
CREATE (d:Dog {name: 'Johnny'})-[l:Loves {unconditionally: true}]->(h:Human)
CREATE (c:Dog {name: 'Floki'});
```

```

> Let's use `merge.relationship` to add a property called `trueLove` to every `Loves` relationship between a human and a dog that has the property unconditional set to true. Otherwise create a new `Loves` relationship with the property unconditionally set to false.

Reviewer: ```suggestion

Use `merge.relationship` procedure to add a property called `trueLove` to every `Loves` relationship between a human and a dog that has the property unconditional set to `true`. Otherwise, create a new `Loves` relationship with the property unconditionally set to `false`.
```

```cypher

> MATCH (d:Dog) MATCH (h:Human) CALL merge.relationship(d, 'Loves', {unconditionally: true}, {unconditionally: false}, h, {trueLove: true}) YIELD rel;

Reviewer: ```suggestion
MATCH (d:Dog)
MATCH (h:Human) 
CALL merge.relationship(d, 'Loves', {unconditionally: true}, {unconditionally: false}, h, {trueLove: true}) 
YIELD rel;
```

</Callout>

### `node()`

The function provides a more flexible way of merging nodes than Cypher’s `MERGE` clause, by
allowing labels and properties to be extracted as results of other procedures
and sent as a list or map.

{<h4> Input: </h4>}

- `labels: List[string]` ➡ A list of all labels that need to be added to the new node.
- `identProps: Map` ➡ A map with key-value pairs used for identifying which nodes will be merged.
- `createProps: Map` ➡ A map with key-value pairs which represent properties of the node, when the procedure creates a new node.
Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
- `mergeProps: Map` ➡ A map with key-value pairs which represent properties of the node, when the procedure merges onto a node.
Equivalent to using `ON MATCH SET` for properties when using `MERGE`.

{<h4> Output: </h4>}

- `node: Node` ➡ The node which was created or merged in the procedure.

{<h4> Usage: </h4>}

Consider a graph with two `Human`, and one `Alien` node.

```cypher
CREATE (h:Human {name:"Mark", age: 20}), (h2:Human {name:"Eva", age: 30}), (a:Alien {name: "Gorlock", age: 20});
```

> Use the `merge.node` procedure on all nodes, and add `galaxy` property on the existing nodes. By leaving `labels` and `identProp` empty,

Reviewer: ```suggestion
Use the `merge.node()` procedure on all nodes, and add `galaxy` property on the existing nodes. By leaving `labels` and `identProp` empty,
```

Should we add () after procedures like I did here?

Author: Maybe we should, yeah

and sent as a list or map.

{<h4> Input: </h4>}

- `startNode: Node` ➡ The start node of the relationship.
- `relationshipType: string` ➡ The relationship type to be matched or created.
- `identProps: Map` ➡ A map with key-value pairs used for identifying which relationships will be merged.
- `createProps: Map` ➡ A map with key-value pairs which represent properties of the relationship, when the procedure creates a new relationship.
Equivalent to using `ON CREATE SET` for properties when using `MERGE`.
- `endNode: Node` ➡ The end node of the relationship.
- `mergeProps: Map` ➡ A map with key-value pairs which represent properties of the relationship, when the procedure merges onto a relationship.
Equivalent to using `ON MATCH SET` for properties when using `MERGE`.

{<h4> Output: </h4>}

- `rel: Relationship` ➡ The relationship which was created or merged in the procedure.

{<h4> Usage: </h4>}

Consider a graph with one `Human`, two `Dog` nodes. Both dogs love the human but one loves him unconditionally.

```cypher
CREATE (d:Dog {name: 'Johnny'})-[l:Loves {unconditionally: true}]->(h:Human)
CREATE (c:Dog {name: 'Floki'});
```

> Use `merge.relationship` procedure to add a property called `trueLove` to every `Loves` relationship between a human and a dog that has the property unconditional set to `true`. Otherwise, create a new `Loves` relationship with the property unconditionally set to `false`.

Reviewer: ```suggestion
Use `merge.relationship()` procedure to add a property called `trueLove` to every `Loves` relationship between a human and a dog that has the property unconditional set to `true`. Otherwise, create a new `Loves` relationship with the property unconditionally set to `false`.
```
